Relocatable OCaml - to opam and beyond

dra27 opened the pull request:
This fourth PR of the trilogy forms the epilogue of Relocatable OCaml, providing the support to allow opam to capitalise on the new features to be able to clone an existing compiler whose configuration matches that requested by the user for a new opam switch. Note that this PR is entirely independent of the other three (in particular, this [epilogue](https://en.wikipedia.org/wiki/Zaphod_Beeblebrox) can in fact be merged before the other three PRs).

Fundamentally, this changes the various `install` targets to consist of a series of macro calls instead of commands. In normal use, these translate to the same commands as before (i.e. `make install` still does the same thing), but the meta-programming of these targets then allows alternative installation mechanisms to be fairly trivially implemented.

Commit-by-commit, to begin with:
- The first four commits deal with various instances where we currently files more than once. While only trivial wasteful at the moment, it's problematic for generating opam files, where a _destination_ file can only be specified once.
- The `install` recipe for `ocamldoc` uses the `$(OCAMLDOC)` and `$(OCAMLDOC_OPT)` variables to refer to the binaries. These are also used as part of the mechanism for selecting the "best" `ocamldoc` (so that the generation of manpages and so forth uses `ocamldoc.opt` if available) but this means that the name is prefixed with `$(ROOTDIR)`. The effect is that instead of installing `ocamldoc/ocamldoc[.exe]` as we do for _all_ other files, we install `./ocamldoc/ocamldoc[.exe]`. The `./` creates problems in an opam `.install`, which expects implicit paths only.
- Currently, files are installed simply by invoking the `install` command. In order to indirect this via a macro call, shell `for` loops can't be used, so there's a bit of necessary GNU make glue to allow the install recipe to be written in bits (`install::`; the same as `partialclean`) and then the `for` loops which were used to install the various binaries in `tools` can be converted to a `$(foreach ...)` which expands to explicit calls. It's noisy, but a mostly mechanical change...
- In the same vein, shell `if` blocks are similarly hairy, and it is somewhat easier to reason about the branching if the macros are responsible for _all_ the logic and the shell simply runs commands. The concern here is that a make macro executes, making a persistent decision (for example to create a directory) which ends up nested in a shell `if` whose conditional is not activated.
- We have patterns everywhere for the installation of artefacts - the `COMPILER_ARTEFACT_DIRS` macro tries to reduce the size of them a little bit, and make it slightly easier to maintain consistency (cf. the typo which had existed, albeit in a rarely used target...)
- `INSTALL_STRIPPED_BYTE_PROG` gets tweaked slightly not to delete the file each time (it still always creates it) so that a `.install` file can refer to it
- The "Straighten out the..." commit consolidates the various variables used to represent installation names (I expect this was already planned once the last few remaining Makefiles were merged). In the same vein, the unmerged Makefiles which also install things now have `SUBDIR_NAME` to go with `ROOTDIR` - i.e. `$(ROOTDIR)/$(SUBDIR_NAME)`

Finally, with everything somewhat more uniform, the commands used for installation are replaced with one of two macro calls. Each macro receives the section it installs to and optionally a subdirectory within that section. The four sections `bin`, `doc`, `lib` and `man` correspond to `--bindir`, `--docdir`, `--libdir` and `--mandir` as passed to `configure`. `libexec` also installs to `--libdir`, but marks the file as executable (`.so` files, `expunge`, etc.). Finally, `stublibs` is an "alias" for the `stublibs` subdirectory, since `opam` has a specific installation section for these.
- `mkdir '$(LIBDIR)/compiler-libs$'` is no longer necessary
-  `install *.cmi '$(LIBDIR)/compiler-libs'` becomes `$(call INSTALL_ITEMS, *.cmi, lib, compiler-libs)`
- Items which need renaming and for which symlinks are required use `INSTALL_ITEM` which takes a _single_ source file, a destination directory. The file can be installed with a different name and a list of basenames for symlinks _in the same directory_ can be specified. For example, the installation `tools/ocamlobjinfo` in a bytecode-only build becomes `$(call INSTALL_ITEM, tools/ocamlobjinfo, bin, , ocamlobjinfo.byte, ocamlobjinfo)` which installs the executable itself as `ocamlobjinfo.byte` and then creates a symlink `ocamlobjinfo` linking to it.
- `INSTALL_BEGIN` and `INSTALL_END` are called, unsurprisingly, at the beginning and end of the `install` recipe and they are called exactly once.

The macro `INSTALL_MODE` controls how `make install` will behave. It defaults to `install`. For each supported value `foo` for `INSTALL_MODE`, there is then a set of macros `INSTALL_DESPATCH_foo_MKDIR`, `INSTALL_DESPATCH_foo_ITEM`, etc., for the 4 installation macros. The macro machinery for `INSTALL_ITEM`, etc., allow a slightly less unnatural use of spaces (in GNU make, `$(call FOO, a)` and `$(call FOO,a)` are not the same call) and also for arguments to be omitted without triggering undefined variable warnings. `_DESPATCH_` macros are then able to written slightly less unnaturally.

5 installation modes are implemented:
- `install` (the default) calls `mkdir`, `install` and `ln`/`cp`, as before
- `display` is a "dry-run" version of `install`
- `list` lists each target file in its installed location (expanding wildcards). opam install files cannot contain duplicate targets, so this mode was useful in identifying some of the earlier de-duplication commits
- `opam` uses the `OPAM_PACKAGE_NAME` variable (which defaults to `ocaml-compiler`) and writes `$(OPAM_PACKAGE_NAME).install` and `$(OPAM_PACKAGE_NAME)-fixup.sh`. The shell script deals with two shortcomings (from the compiler's perspective) of opam install files. The `doc` section in the install file would install to `doc/ocaml-compiler`, rather than `doc/ocaml` and there currently no way to override this. The install format also cannot copy or create symlinks. `ocaml-variants.opam` in the repo is updated to use this mechanism, but I don't suggest we deploy it to the release packages until a future version of opam deals with these two limitations (or we do something which means they're no longer an issue)
- `clone` write a single `$(OPAM_PACKAGE_NAME)-clone.sh` script. This shell script is designed to be run from the installation prefix and clones the compiler to the single argument argument it's passed. This is the key mechanism needed for Relocatable OCaml, as installing this script into opam switches with the compiler allows that compiler to be cloned to another switch.

The recommended workflow for installing a built compiler was inadvertently broken in #13160. The underlying problem is tracked in ocaml/opam#6621 but two commits which workaround this are added which then allows the compiler built in GitHub Actions to be tested as an opam switch (in addition to providing CI for this somewhat advanced workflow, it also means that the opam file is tested in CI without having to build the compiler again).

dra27 committed:
Eliminate local for loops in install

Continue to use for when invoking recursive make calls, but use
$(foreach ..) to generate command sequences so that macros can be used
in them.

dra27 committed:
Use macros to generate installation commands

make install works as it did before. All the commands in the install
targets now go through a macro call which allows the semantic intent of
each command to be more clearly specified.

dra27 committed:
Add additional modes to make install

make [INSTALL_MODE=install] install
  - installs the compiler, as normal
make INSTALL_MODE=display install
  - displays the operations needed for
make INSTALL_MODE=list install
  - lists the files and symbolic links which are installed

dra27 committed:
Don't explicitly install toplevel/byte/*.cmi

All of the .mli files in toplevel/byte are "common", so they're already
installed by the patterns in toplevel.

dra27 committed:
Don't explicitly install main.cmx and optmain.cmx

It's necessary to install main.o and optmain.o because the two modules
are not part of ocamlbytecomp.cmxa and ocamloptcomp.cmxa, but the .cmx
files are already installed as part of wildcard patterns on driver/

dra27 committed:
Remove duplicate installation in installopt

The ocamldoc installation commands in installopt duplicate a series of
commands which already happen in install.

dra27 committed:
Don't install ocamldoc's .cmi files twice

dra27 committed:
Use implicit names when installing ocamldoc

The ocamldoc binaries were installed using a relative path. They're the
only binaries installed that way - switch them to use an implicit path,
as it's easier to make an implicit path subsequently explicitly relative
if needed than vice versa.

dra27 committed:
Generate an opam .install file and support script

make INSTALL_MODE=opam install generates $OPAM_PACKAGE_NAME.install and
$OPAM_PACKAGE_NAME-fixup.sh ($OPAM_PACKAGE_NAME defaults to
ocaml-compiler). Nothing is installed by this mode. The fixup.sh script
is intentionally not made executable (it should be invoked explicitly
with sh) and creates symbolic links, if required, and also manually
copies the files to the doc dir, as the .install file format doesn't
allow the correct location to be specified.

dra27 committed:
Experimentally, use it

dra27 committed:
Keep the stripped bytecode binaries

No need to remove things during install when it will be done later by a
package manager. Also prepares for the possibility of letting something
else install the binary subsequently.

dra27 committed:
Add SUBDIR_NAME to principal Makefiles

SUBDIR_NAME goes along with ROOTDIR - where ROOTDIR is the relative path
from Makefile back to the root directory, SUBDIR_NAME is the implicit
path from the root directory to Makefile.

dra27 committed:
Straighten out the INSTALL_* Makefile variables

Move various definitions around such that merged Makefile variables are
all in Makefile.common. Sanitise the names to be closer to
INSTALL_LIBDIR_subdirname.

dra27 committed:
Generate a relocatable cloning script

dra27 committed:
Sprinkle some meta-programming on the install target

dra27 committed:
Remove invalid directory in source install

Line was added (with typo) in PR#8925 (4.09) and should have been
removed as part of PR#10355 (4.13)

dra27 committed:
Mark winpthreads dependency as Windows-only

Redundant filter, but mitigates an issue with --assume-built in opam

dra27 committed:
Mark input dependencies as build

Mitigates an issue with opam install --assume-built since build
dependencies are ignored. The semantics should remain consistent: in
particular, as there is only a single version of each of these packages,
the key issue is that removing the package will still trigger the
correct behaviour as the dependency graph will change.

dra27 committed:
Test opam in CI

dra27 committed:
Eliminate local if blocks in install

dra27 committed:
Got an interpreter, and we're not afraid to use it

dra27 committed:
Remove the recursive invocation in install targets

shym commented on Makefile:
```diff
@@ -2682,319 +2680,302 @@ endif
+
+# These three targets are the slightly esoteric special sauce that avoid
+# recursive make invocations in the install targets.
+# There are three basic install recipies:
```
```suggestion
# There are three basic install recipes:
```

dra27 replied:
🥧

shym commented on Makefile:
```diff
@@ -2682,319 +2680,302 @@ endif
+#   native-install
+# - The old installoptopt target is also available as full-installoptopt and
+#   installopt
+# These sets of recipies are then welded together by these three dependency
```
```suggestion
# These sets of recipes are then welded together by these three dependency
```

dra27 replied:
Yeah, I appear to have a spelling blind spot there - consistently wrong, at least 🤷

shym commented on Makefile:
```diff
@@ -2682,319 +2680,302 @@ endif
+#   installopt
+# These sets of recipies are then welded together by these three dependency
+# specifications
+# - When configured with --disable-native-compiler, the install target simply
+#   depends on common-install (see above)
```
Would it make sense to deprecate explicitly the old targets?

dra27 replied:
I was shying away from that, just because I expected that to happen ultimately as part of the final stages of the `Makefile` merging project

shym commented on Makefile.common:
```diff
@@ -83,15 +83,334 @@ V_ODOC       =
+
+INSTALL_ENSURE_DIR = \
+  $(if $(filter undefined,$(origin DIR_CREATED_$(subst exec,,$(1))_$(2))),$\
+    $(eval DIR_CREATED_$(1)_$(2):=)$\
```
```suggestion
    $(eval DIR_CREATED_$(subst exec,,$(1))_$(2):=)$\
```
IIUC

dra27 replied:
Yes, indeed, thanks - there should presumably be a stray `mkdir` in the install instructions

shym commented on Makefile.common:
```diff
@@ -83,15 +83,334 @@ V_ODOC       =
+    $(SH_AND)
+
+MK_LINK = \
+  (cd "$(INSTALL_SECTION_$(2))$(addprefix /,$(3))" && \
```
Why not use `QUOTE_SINGLE` here also?

dra27 replied:
Hmm, it's not necessarily the most acceptable of explanations, but my reasoning was only to be using `QUOTE_SINGLE` where I _can't_ be certain that single quotes _won't_ appear (I may not have been consistent, though, given that the two arguments being linked are being passed to `QUOTE_SINGLE`). The rationale is that `cmd` makes all of this a bit of a nightmare anyway... i.e. the build system itself should _never_ introduce something as a stupid as an apostrophe... but we must cope with the fact that a user path may quite legitimately include one! (I mean that while I wouldn't, I think it should be possible to install to `C:\David's Work`, but I think we would be very silly to engineer `ocamlopt's finest binary.exe` as one of our filenames!)

shym commented on Makefile.common:
```diff
@@ -83,15 +83,334 @@ V_ODOC       =
+#   $2 = subdirectory (may be empty, but otherwise must end with "/")
+#   $3 = name to install $1 (must be specified)
+# The compiler is installed as the ocaml package in opam, but the actual files
+# are installed from other packages (typically ocaml-compiler). For the the lib
```
```suggestion
# are installed from other packages (typically ocaml-compiler). For the lib
```

shym commented on tools/opam/generate.ml:
```diff
@@ -0,0 +1,221 @@
+
+let output_endline oc = Printf.kfprintf (fun oc -> output_char oc '\n') oc
+
+(* [generate_file file] processes then erases opam-bin, opam-lib opam-libexec
```
```suggestion
(* [generate_install file] processes then erases opam-bin, opam-lib opam-libexec
```
I suppose

shym commented on tools/opam/generate.ml:
```diff
@@ -0,0 +1,221 @@
+    else begin
+      let mklink (dir, target, source) =
+        (* Convert all slashes to _two_ backslashes *)
+        let to_backslashes oc s =
+          output_string oc (String.concat {|\\|} (String.split_on_char '/' s))
+        in
+        output_endline oc
+          {|  cmd /c "mklink %a\\%s %s"|} to_backslashes dir source target
```
No `Filename.quote` is needed there? (That might be harder, because it will end up in string, sigh)

dra27 replied:
The quoting here is absolutely hellish - the problem is that `mklink` is not an executable. I went on the basis that the user who decides to do `./configure --host=x86_64-w64-mingw32 --bindir 'C:\OCaml\I like bin's\'` can flippin' well contribute the patch to quote it 😂

shym reviewed (commented):
I read through this PR, but not testing it (yet). I like very much the cleanup in the installation rules!
I’ve taken a few notes, here and inlined.

- In a0933f257884569bcb6c36b207bed743fa405ef0, installation of native DLLs is removed: was that on purpose?
- In 79c7fea655a6611b44eea399a7fdb618fc368e10 commit message, “relative” means “explicit”, I guess?
- 515e861774ea757be7ab725ff5141b409dbc53da creates a very elegant solution, congratulations :-)
- 7fee32ca495d9afe948ca781a45d3f4d8922b22e: what a relief! :-)

MisterDA commented on .github/workflows/build-msvc.yml:
```diff
@@ -193,8 +202,24 @@ jobs:
+          opam pin add --no-action winpthreads winpthreads
+          opam install --yes flexdll winpthreads
+          opam install --yes --assume-built ocaml-variants
+          opam exec -- ocamlc -v
```
I suppose that the opam cli is quite stable at this point; but we could pass `--cli=2.4` or set `OPAMCLI` for future-proofing.

dra27 replied:
I should indeed use the features of `opam` that I originally designed/specified and recommend others use!! 🫣😂

MisterDA commented on tools/ci/appveyor/appveyor_build.cmd:
```diff
@@ -103,6 +114,9 @@ if "%BOOTSTRAP_FLEXDLL%" equ "true" (
 cd "%APPVEYOR_BUILD_FOLDER%"
 appveyor DownloadFile "https://github.com/ocaml/flexdll/archive/%FLEXDLL_VERSION%.tar.gz" -FileName "flexdll.tar.gz" || exit /b 1
 appveyor DownloadFile "https://github.com/ocaml/flexdll/releases/download/%FLEXDLL_VERSION%/flexdll-bin-%FLEXDLL_VERSION%.zip" -FileName "flexdll.zip" || exit /b 1
+appveyor DownloadFile "https://github.com/ocaml/opam/releases/download/2.4.1/opam-2.4.1-x86_64-windows.exe" -FileName "opam.exe" || exit /b 1
```
You could introduce `OPAM_VERSION` somewhere.

dra27 replied:
To what end, though - it's used exactly once here? I'm never sure about polluting the environment in these horrible batch files - `FLEXDLL_VERSION` has to be communicated between three different files, but this one's localised.

shym reviewed (commented):
I tested the installation of a cross compiler, using both the standard and the opam modes and it worked as hoped, congratulations!
I could trim down a little bit the dummy files that are created just to reuse the common installation workflow:
```diff
diff --git i/Makefile.cross w/Makefile.cross
index b0a21c3d7c..362eb15aa9 100644
--- i/Makefile.cross
+++ w/Makefile.cross
@@ -105,11 +105,7 @@ INSTALL_OVERRIDES=build_ocamldoc=false WITH_DEBUGGER= OCAMLRUN=ocamlrun
 .PHONY: installcross
 installcross:
        # Create dummy files to keep `install` happy
-       touch \
-         $(addprefix toplevel/, \
-           $(foreach ext,cmi cmt cmti cmx, native/nat__dummy__.$(ext)) \
-             all__dummy__.cmx topstart.o native/tophooks.cmi)
+       touch toplevel/topstart.o
        $(LN) `command -v ocamllex` lex/ocamllex.opt$(EXE)
-       $(LN) `command -v ocamlyacc` yacc/ocamlyacc.opt$(EXE)
        # Real installation
        $(MAKE) install $(INSTALL_OVERRIDES)
```

dra27 commented:
> I read through this PR, but not testing it (yet). I like very much the cleanup in the installation rules! I’ve taken a few notes, here and inlined.
> 
> * In [a0933f2](https://github.com/ocaml/ocaml/commit/a0933f257884569bcb6c36b207bed743fa405ef0), installation of native DLLs is removed: was that on purpose?

Oo, it's intentional (they should never have been being installed in the first place), but I'm not sure I should be doing it so subtly, indeed! It was a misunderstanding that crept in with the tsan changes, I think - we never need stub DLLs for native code (because .cmxs is already a stub DLL...)

> * In [79c7fea](https://github.com/ocaml/ocaml/commit/79c7fea655a6611b44eea399a7fdb618fc368e10) commit message, “relative” means “explicit”, I guess?

Yes - I should be explicit about explicitness, indeed, given that all the paths are relative!!

> * [515e861](https://github.com/ocaml/ocaml/commit/515e861774ea757be7ab725ff5141b409dbc53da) creates a very elegant solution, congratulations :-)
> * [7fee32c](https://github.com/ocaml/ocaml/commit/7fee32ca495d9afe948ca781a45d3f4d8922b22e): what a relief! :-)

Thank you 🙂

shym commented:
> > * In [a0933f2](https://github.com/ocaml/ocaml/commit/a0933f257884569bcb6c36b207bed743fa405ef0), installation of native DLLs is removed: was that on purpose?
> 
> Oo, it's intentional (they should never have been being installed in the first place), but I'm not sure I should be doing it so subtly, indeed! It was a misunderstanding that crept in with the tsan changes, I think - we never need stub DLLs for native code (because .cmxs is already a stub DLL...)

I thought it strange for them to be dropped without a mention in the commit message, even if I didn’t know when (that is to say, even just if) they were useful.

Maybe we could drop the `byt` in `dll*byt*` filenames (if that doesn’t break backward compatibility, though)?

dra27 commented:
Rebased - review responses to follow

shym commented on tools/opam/generate.ml:
```diff
@@ -0,0 +1,221 @@
+    Sys.remove file
+  end
+
+(* [generate_install file] processes then erases opam-bin, opam-lib opam-libexec
+   and opam-man to produce [file] *)
```
```suggestion
(* [generate_install file] processes then erases opam-bin, opam-lib,
   opam-libexec and opam-man to produce [file] *)
```

shym reviewed (approved):
I went through the `range-diff` and the “Review” commits (thank you so much yet again for such a clean update!): I managed to find only a missing comma :-D So this looks good to me!

damiendoligez reviewed (approved):
Approved after interactive review; still some final testing to do.

dra27 commented:
Rebased on to #14245

kit-ty-kate commented on tools/opam/process.sh:
```diff
@@ -0,0 +1,182 @@
+#*                                                                        *
+#**************************************************************************
+
+set -e
```
for some extra safety maybe?
```suggestion
set -eu
if set -o pipefail > /dev/null 2> /dev/null; then
  # POSIX.1-2024
  set -o pipefail
fi
```

kit-ty-kate replied:
same question for the generated `-clone` and `-fixup` scripts

dra27 replied:
Yes, indeed - that's a good idea!

dra27 replied:
What am I missing in calling it twice? Isn't

```sh
set -o pipefail > /dev/null 2> /dev/null || true
```

sufficient?

kit-ty-kate replied:
oh yes, of course

dra27 replied:
Hah - it turns out neither of us appears to have been correct! Apparently shell builtins are allowed to abort execution unconditionally, so it appears to be necessary to have:

```sh
if (set -o pipefail 2>/dev/null); then
  set -o pipefail
fi
```

With Ubuntu's default dash, both this:

```sh
#!/bin/sh

set -o pipefail || echo No pipefail
```
and
```sh
#!/bin/sh

if ! set -o pipefail; then
  echo No pipefail
fi
```
display an error from the `set` command but then immediately abort the script.

kit-ty-kate commented on tools/opam/generate.ml:
```diff
@@ -0,0 +1,221 @@
+    let _ =
+      let create_dir seen (dir, _, _) =
+        if not (StringSet.mem dir seen) && String.contains dir '/' then
+          output_endline oc {|mkdir -p '%s'|} dir;
```
It's probably best to avoid the silent invariant that no `'` character exists in `dir`
```suggestion
          output_endline oc {|mkdir -p %s|} (Filename.quote dir);
```

dra27 replied:
That just replaces it with a different invariant on Windows (`Filename.quote` doesn't do the same thing). I think it's more hassle than it's worth, given that this structure is _only_ the root directories of opam switches (`lib/ocaml`, `bin`, etc.)

kit-ty-kate commented on tools/opam/generate.ml:
```diff
@@ -0,0 +1,221 @@
+        String.map (function '@' -> '/' | c -> c)
+                   (String.sub file 6 (String.length file - 6))
+      in
+      output_endline oc {|mkdir -p "$1/%s"|} dir;
```
```suggestion
      output_endline oc {|mkdir -p "$1"/%s|} (Filename.quote dir);
```

kit-ty-kate commented on tools/opam/generate.ml:
```diff
@@ -0,0 +1,221 @@
+  In_channel.fold_lines (fun _ line ->
+    match String.split_on_char ' ' line with
+    | [source; dest] ->
+        output_endline oc {|cp '%s' "$1/%s/%s"|} source dir dest
```
```suggestion
        output_endline oc {|cp %s "$1"/%s/%s|}
          (Filename.quote source) (Filename.quote dir) (Filename.quote dest)
```

kit-ty-kate commented on tools/opam/process.sh:
```diff
@@ -0,0 +1,182 @@
+  if [ -e "$OPAM_SWITCH_PREFIX/share/ocaml/build-id" ]; then
+    switch="$(tail -n 1 "$OPAM_SWITCH_PREFIX/share/ocaml/build-id")"
+    if [ -n "$switch" ]; then
+      switch_share_dir="$(opam var --switch="$switch" share | tr -d '\r')"
```
```suggestion
      switch_share_dir="$(opam var --safe --switch="$switch" share | tr -d '\r')"
```

kit-ty-kate commented on tools/opam/process.sh:
```diff
@@ -0,0 +1,182 @@
+  fi
+
+  echo "🐫 Requesting list of switches from opam"
+  opam switch list --short | tr -d '\r' | grep -Fxv "$OPAMSWITCH" \
```
```suggestion
  opam switch list --safe --short | tr -d '\r' | grep -Fxv "$OPAMSWITCH" \
```

kit-ty-kate commented on tools/opam/process.sh:
```diff
@@ -0,0 +1,182 @@
+    >> opam-switches 2> /dev/null || true
+
+  while IFS= read -r switch; do
+    switch_share_dir="$(opam var --switch="$switch" share | tr -d '\r')"
```
```suggestion
    switch_share_dir="$(opam var --safe --switch="$switch" share | tr -d '\r')"
```

kit-ty-kate commented on tools/opam/process.sh:
```diff
@@ -0,0 +1,182 @@
+    "$make" install
+  else
+    origin="$(tail -n 1 build-id)"
+    origin_prefix="$(opam var --switch="$origin" prefix | tr -d '\r')"
```
```suggestion
    origin_prefix="$(opam var --safe --switch="$origin" prefix | tr -d '\r')"
```

kit-ty-kate commented on tools/opam/generate.ml:
```diff
@@ -0,0 +1,221 @@
+
+let clone_files oc dir ic =
+  output_endline oc
+    {|dest="$1/%s" xargs sh "$1/clone-files" <<'EOF'|} dir;
```
```suggestion
    {|dest="$1"/%s xargs sh "$1/clone-files" <<'EOF'|} (Filename.quote dir);
```

kit-ty-kate reviewed (commented):
I only took a couple of hours to briefly look at `tools/opam/*` but the general idea looks good, albeit a bit complex to decipher at first between all the various scripts and the [planned opam-repository changes](https://github.com/dra27/opam-repository/blob/e2babed23231596afefc2e346609f07719859386/packages/relocatable-compiler/relocatable-compiler.5.5.0.20251109.1/opam). I only had a couple of minor remarks that can be ignored if i was off (the `opam --safe/readonly` bits should probably be added though).

Hopefully opam will be able to clone relocatable packages itself in the future, which should allow to shave off a bit of the complexity by removing the clone mode, but that will probably need some time to be designed and implemented correctly.

Also the addition of the .install file is a really nice improvement i think.

dra27 commented:
I like the `--safe` idea - but I'd ruled out `Filename.quote` - as the quoting stands, it should be exactly quoting actual untrusted input the `"$1"`, etc. while leaving names which we are entirely in control of on the build/install side not at risk of being misquoted by an unexpected platform interaction!

dra27 commented:
Thanks, @kit-ty-kate! `--safe` added to those calls; in re-checking the equivalence of the interim shell scripts and `generate.ml`, I spotted that the `$CP` fallbacks weren't being correctly escaped on Windows and there were a couple of unnecessary differences between the two versions.

I haven't changed the `set` mode on `process.sh` _yet_, just because I want to plug that into a test properly, and the "apparatus" for doing that is out-of-sync at the moment!

dra27 commented:
Separately, this has been tortured on [precheck#1091](https://ci.inria.fr/ocaml/job/precheck/1091/) with a special check that builds the compiler and the trunk commit it sits on and compares the installation artefacts (i.e. checks that the same files are installed) and then repeats the entire process for `--disable-suffixed`. It seems to be passing!

dra27 committed:
Explicitly enforce filename invariants

Additional hardening on the requirements for names of files and
directories in the distribution.

dra27 commented:
Updated torture version going through [precheck#1092](https://ci.inria.fr/ocaml/job/precheck/1092/). I've added a commit at the end of the series which adds guards in the Makefile-machinery to detect filenames which would cause problem, I've also - for the sake of history, and because it was a useful extra check - ensured that the shell and OCaml generators are actually producing the same scripts and hardened the use of single-quoting on all-generated filenames.

shym commented on Makefile.common:
```diff
@@ -141,13 +175,27 @@ INSTALL_ITEM = \
+# $(1) can be a mix of filenames and globs. Check $(1) itself for illegal
+# characters (i.e. ensure that the whole of $(1) is a "valid" path) and then for
+# each word containing a pattern character also evaluate the pattern and check
+# the resulting files. $(2)/$(3) be be a valid section.
```
`be be` → `must be` maybe?

dra27 commented:
Last couple of tweaking notes to those macros added. However, @shym and I discussed this separately - the benefit of the make macros is that it would catch errors just doing a normal `make install`, but that's not a normal part of a development workflow - given that CI is always going to be checking the opam/clone mechanisms, it therefore seems sufficient to keep the sanity checks in tools/opam/generate.ml and remove at least some of the meta-programming in Makefile.common!

dra27 commented:
Assuming everything passes CI again here, and everyone remains happy, I think this is good to go, too. Relocatable and now easy to clone, too...!

dra27 commented:
This happened on #14244 as well - the AppVeyor issue is that the "CI: Full matrix" check is occasionally pushing this over its 1 hour cancellation mark on the occasions when we get an over-stretched worker.
